README.md
[![Circle CI][circle-image]][circle-url]

# DEPRECATED COMPONENT. USE https://github.com/elasticio/rest-api-component-v2 instead

# REST API component

The **REST API component** is a simple yet powerful component that allows you to connect to any REST API without programming your own components and deploying them into the platform.

The REST API component will perform a single REST API call when executed. Incoming data can gets used to configure the API call made and the response from the API call will be the output.

This document covers the following topics:

*   [Introduction](#introduction)
*   [Authorisation methods](#authorisation-methods)
*   [Defining HTTP headers](#defining-http-headers)
*   [Defining request body](#defining-request-body)
*   [Working with XML Response](#working-with-xml)
*   [HTTP Headers in Response](#http-headers)
*   [Redirection](#redirection)
*   [Attachments](#attachments)
*   [Exception handling](#exception-handling)
*   [Known Limitations](#known-limitations)

## Introduction

The example below shows the development team creation using the REST API component with our own [REST API service](https://api.elastic.io/docs "elastic.io REST API service").

![image](https://user-images.githubusercontent.com/22715422/87129437-000fa980-c29a-11ea-920c-cc161db6cb3a.png)
*Numbers show: (1) The URL and method of the REST API resource, (2) the HTTP call headers. (3) configuration options and (4) follow redirect mode.*

1.  HTTP methods and URL
 * REST API component supports the following HTTP methods: `GET`, `PUT`, `POST`, `DELETE` and `PATCH`.
 * The URL of the REST API resources. Accepts JSONata expressions, meaning the URL address evaluates [JSONata](http://jsonata.org/) expressions.
2. Request Headers and Body
 * Definition of request [headers](#defining-http-headers)
 * Definition of request [body](#defining-http-body), if the HTTP method is not `GET`
3. Configuration options
 * ``Don`t throw Error on Failed Calls`` - if enabled return error, error code and stacktrace in message body otherwise throw error in flow.
 * ``Split Result if it is an Array`` - if enabled and response is array, creates message for each item of array. Otherwise create one message with response array.  
 * ``Retry on failure`` - enabling [rebound](https://support.elastic.io/support/solutions/articles/14000044750-why-and-where-we-use-the-rebound-) feature for following HTTP status codes:
    - 408: Request Timeout
    - 423: Locked
    - 429: Too Many Requests
    - 500: Internal Server Error
    - 502: Bad Gateway
    - 503: Service Unavailable
    - 504: Gateway Timeout
    - DNS lookup timeout
4. ``Do not verify SSL certificate (unsafe)`` - disable verifying the server certificate - **unsafe**.
5. ``Follow redirect mode`` - If you want disable Follow Redirect functionality, you can use option ``Follow redirect mode``.By default ``Follow redirect mode`` option has value ``Follow redirects``.
6. ``Delay`` - If you want to slow down requests to your API you can set delay value (in seconds) and the component will delay calling the next request after the previous request.
Time for the delay is calculated as `Delay`/ `Call Count` and shouldn't be more than 1140 seconds (19 minutes due to platform limitation). 
The `Call Count` value by default is 1. If you want to use another value, please set the `Call Count` field. 
Notice: See [Known Limitations](#known-limitations) about `Delay` value.
7. ``Call Count`` - the field should be used only in pair with `Delay`, default to 1.
8. ``Request timeout`` - Timeout period in milliseconds (1-1140000) while component waiting for server response, also can be configured with REQUEST_TIMEOUT environment variable if configuration field is not provided. Defaults to 100000 (100 sec).
Notice: Specified for component REQUEST_TIMEOUT enviroment variable would be overwritten by specified value of Request timeout, default value would be also overwritten

## Authorisation methods

To use the REST API component with any restricted access API provide the authorisation information.

![alt text](https://cdn.elastic.io/documentation/restapi-component-auth.png "REST API component Basic authorisation")
*Example above shows how to add the username/password to access the API during the integration flow design.*

You can add the authorisation methods during the integration flow design or by going to your `Settings > Security credentials > REST client` and adding there.

REST API component supports 4 authorisation types:

*   `No Auth` - use this method to work with any open REST API
*   `Basic Auth` - use it to provide login credentials like **username/password**
*   `API Key Auth` - use it to provide `API Key` to access the resource
*   `OAuth2` - use it to provide `Oauth2` credentials to access the resource. Currently it is implemented `Authorization code` OAuth2 flow.

Please note that the result of creating a credential is an HTTP header automatically placed for you. You can also specify the authorisation in the headers section directly.

## Defining HTTP headers

Use this section to add the request headers.

![alt text](https://cdn.elastic.io/documentation/rest-api-component-headers-get.png "REST API component Headers field")

Each header has a name and a value. Header name should be colon-separated name-value pairs in clear-text `string` format. The header value can use [JSONata](http://jsonata.org/) expressions.

*Note:* **HTTP Response headers** will not be stored, the components stores body and attachment only.
 
## Defining request body

The body may be defined if the HTTP method is not `GET`. The **body** tab enables configuration options such as the **content type** drop-down menu and the **body input field**.

Here is the list of all supported **content types**:

*   `multipart/form-data`
*   `application/x-www-form-urlencoded`
*   `text/plain`
*   `application/json`
*   `application/xml`
*   `text/xml`
*   `text/html`

The **body input field** changes according to the chosen content type.

*Notes:* 
1. **Response body** will be stored in msg.body
2. Request body that couses empty response body will return `{}`

### Sending JSON data

Here is how to send a JSON data in the body. Change the **content type** to `application/json` and the **body input part** would change accordingly to accept JSON object. Please note that this field supports [JSONata](http://jsonata.org) expressions.

![alt text](https://cdn.elastic.io/documentation/restapi-component-body-json-var.png "REST API component Body sending JSON data")
*Example shows the JSON in the body where the `name` parameter value gets mapped using the value of `project_name` from the previous step of integration.*

### Sending XML data

To send an `XML` data set the content type to `application/xml` or `text/xml` and place the `XML` in the body input field between double-quotes like:

```
"
<note>
  <to>" & fname & "</to>
  <from>Jani</from>
  <heading>Reminder</heading>
  <body>Don't forget me this weekend!</body>
</note>
"
```
Use a JSONata expression to include and map any values coming from the previous steps. It will replace the variable with a real value in the final mapping. Note that the rest of `XML` gets passed as a `string`.

### Sending Form data

To send a form data two content types are available:

*   `application/x-www-form-urlencoded` - used to submit simple values to a form
*   `multipart/form-data` - used to submit (non-alphanumeric) data or file attachment in payload

In both cases the payload gets transmitted in the message body.

In case of `application/x-www-form-urlencoded` content type add the necessary parameters by giving the name and the values like:

![alt text](https://cdn.elastic.io/documentation/restapi-component-body-form-simple.png "REST API component Body sending a simple form")
*Please note that parameter value fields support [JSONata](http://jsonata.org) expressions.*

This HTTP request would submit `key1=value1&key2=value2` in the message body.

In case of `multipart/form-data` content type add the parameters similarly.

![alt text](https://cdn.elastic.io/documentation/restapi-component-body-form-complex.png "REST API component Body sending a complex form")

The transmitted HTTP request body would be:

```
--__X_ELASTICIO_BOUNDARY__
Content-Disposition: form-data; name="part1"

Please note that this fields supports [JSONata](http://jsonata.org) expressions.
--__X_ELASTICIO_BOUNDARY__
Content-Disposition: form-data; name="part2"

<p>Some more text</p>
--__X_ELASTICIO_BOUNDARY__--
```

Notice how different parts get separated by the boundary. This form is capable of supporting attachments as well.

### Working with XML

This component will try to parse XML content types in the HTTP Response assuming the `Content-Type` header has a
**MIME Content Type** with `xml` in it (e.g. `application/xml`). 
In this case response body will be parsed to JSON using `xml2js` node library and following settings:

```js
{
    trim: false,
    normalize: false,
    explicitArray: false,
    normalizeTags: false,
    attrkey: '_attr',
    tagNameProcessors: [
        (name) => name.replace(':', '-')
    ]
}
```

for more information please see the 
[Documenattion of XML2JS library](https://github.com/Leonidas-from-XIV/node-xml2js#options)

## HTTP Headers 

You can to get HTTP response header only if ``Don`t throw Error on Failed Calls`` option is checked.
In this case output structure of component will be: 
```js
    {
      headers:<HTTP headers>,
      body:<HTTP response body>,
      statusCode:<HTTP response status code>
      statusMessage:<HTTP response status message>
    }
```

## Attachments
Rest API component has opportunity of binary data sending. You just need choose ``multipart/form-data`` Content type and attachments from input message will be included to the request payload automatically.

Rest-api component automatically load binary data to attachments with next content types in response headers:
* image/*
* text/csv
* application/msword
* application/msexcel
* application/pdf
* application/octet-stream
* application/x-binary
* application/binary
* application/macbinary

## Exception handling
Rest API component uses exception handling logic below: 
![Exception handling logic](https://user-images.githubusercontent.com/13310949/41960520-9bd468ca-79f8-11e8-83f4-d9b2096deb6d.png)

## Known Limitations

**1.** The component can parse any of json and xml content types. 
There are:
* application/json
* application/xml
* text/xml
* etc.

`If content type is not  exists  in response header, component will try parse response as json. 
If it get parse exception, it return response as is.`

**2.** Attachments limitations:

    1. Maximal possible size for an attachment is 10 MB.
    2. Attachments mechanism does not work with [Local Agent Installation](https://support.elastic.io/support/solutions/articles/14000076461-announcing-the-local-agent-)

**3.** OAuth2 authentication strategy limitation: [Access Token Response](https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/) contains `refresh_token` optional property, but due to EIO platform limitation it is required.
Possible solution - use access_type:offline in additional parameters (may not work in some cases).

**4.** We suggest not to set Delay value more then time period between two executions of the flow.
Please keep in mind that delay can influence on time of next execution. 
For example, the flow has type `Ordinary` and scheduled to execution for every 1 minute, but the delay is set to 120 sec, so the next execution will be started only after 120 sec, instead of 1 minute.

[circle-image]: https://circleci.com/gh/elasticio/rest-api-component.svg?style=svg&circle-token=2bf8e1f60133011d1fdea9505afdbabbd12b0c7b
[circle-url]: https://circleci.com/gh/elasticio/rest-api-component


CHANGELOG.md
## 1.2.17 (December 7, 2020)
* Update sailor version to 2.6.21

## 1.2.16 (November 10, 2020)

* Change logo to grayscale.

## 1.2.15 (November 6, 2020)
* Update sailor version to 2.6.18

## 1.2.14 (October 22, 2020)
* Annual audit of the component code to check if it exposes a sensitive data in the logs
* Update sailor version to 2.6.17
* Update internal libraries versions

## 1.2.13 (October 8, 2020)
* Component got deprecated. Use https://github.com/elasticio/rest-api-component-v2 instead

## 1.2.12 (September 25, 2020)
* Revert changes to url encoding

## 1.2.11 (September 6, 2020)
* Fix url encoding

## 1.2.10 (August 25, 2020)
* Fix `application/x-www-form-urlencoded` encoding bug

## 1.2.9 (July 10, 2020)
* Timeout configuration field

## 1.2.8 (July 6, 2020)
* Add configuration fields: `Delay` and `Call Count` for rate limit
* Update sailor version to 2.6.13

## 1.2.7 (June 24, 2020)
* Add checkBox to not verify servers certificate

## 1.2.6 (June 6, 2020)
* Remove update docs on deploy script

## 1.2.5 (May 19, 2020)
* Update sailor version to 2.6.7

## 1.2.4 (April 9, 2020)
* Fix `No refresh tokens were returned by the OAuth2 provider` error for credentials with non-expiring refresh_token

## 1.2.3 (April 1, 2020)
* New Jsonata expressions support: getFlowVariables and getPathrough

## 1.2.2 (March 19, 2020)
* Add binary message types to list for processing them as attachments

## 1.2.1(March 12, 2020)

### Added
* Add validity check of `refresh_token` in keys for `OAuth2` authentication strategy

## 1.2.0(February 13, 2020)

### Added
* Added new authentication strategy `OAuth2`

## 1.1.4 (December 23, 2019)

* Update sailor version to 2.5.4
* Remove Enable debugging checkbox

## 1.1.3 (December 07, 2019)

* Update sailor version to 2.5.1
* Update jsonata-moment to 1.1.4

## 1.1.2 (September 06, 2019)

* Fix `new Buffer()` deprecated warning

## 1.1.0 (August 28, 2019)

### Added
* Added feature `Retry on failure`

## 1.0.1 (June 26, 2019)

* Initial release which includes a bunch of previous unversioned releases


lib/utils.js
/* eslint-disable max-len,no-shadow,no-param-reassign,no-underscore-dangle,no-use-before-define,consistent-return,arrow-parens */

const { JsonataTransform } = require('@elastic.io/component-commons-library');
const request = require('request-promise');
const { messages } = require('elasticio-node');
const xml2js = require('xml2js-es6-promise');
const uuidv1 = require('uuid/v1');

const attachment = require('./attachments');
const {
  getRateLimitDelay, rateLimit, getRequestTimeout, encodeWWWFormParam,
} = require('./helpers');

const HTTP_ERROR_CODE_REBOUND = new Set([408, 423, 429, 500, 502, 503, 504]);

const methodsMap = {
  DELETE: 'delete',
  GET: 'get',
  PATCH: 'patch',
  POST: 'post',
  PUT: 'put',
};

const bodyEncodings = {
  FORM_DATA: 'form-data',
  RAW: 'raw',
  URLENCODED: 'urlencoded',
};

const bodyMultipartBoundary = '__X_ELASTICIO_BOUNDARY__';

const contentTypes = {
  FORM_DATA: 'multipart/form-data',
  URLENCODED: 'application/x-www-form-urlencoded',
  TEXT: 'text/plain',
  APP_JSON: 'application/json',
  APP_XML: 'application/xml',
  TEXT_XML: 'text/xml',
  HTML: 'text/html',
};

const formattedFormDataHeader = `multipart/form-data; charset=utf8; boundary=${bodyMultipartBoundary}`;

const authTypes = {
  NO_AUTH: 'No Auth',
  BASIC: 'Basic Auth',
  API_KEY: 'API Key Auth',
  OAUTH2: 'OAuth2',
};

const CREDS_HEADER_TYPE = 'CREDS_HEADER_TYPE';

/**
 * Executes the action's/trigger's logic by sending a request to the assigned URL and emitting response to the platform.
 * The function returns a Promise sending a request and resolving the response as platform message.
 *
 * @param {Object} msg incoming messages which is empty for triggers
 * @param {Object} cfg object to retrieve triggers configuration values, such as, for example, url and userId
 * @returns {Object} promise resolving a message to be emitted to the platform
 */
/* eslint-disable-next-line func-names */
module.exports.processMethod = async function (msg, cfg) {
  const emitter = this;

  const config = cfg.reader;

  if (!config.url) {
    throw new Error('URL is required');
  }
  const url = JsonataTransform.jsonataTransform(msg, { expression: config.url }, emitter);
  const { method, headers } = config;
  const body = config.body || {};
  const followRedirect = cfg.followRedirect !== 'doNotFollowRedirects';
  const { auth } = cfg;
  const requestTimeout = getRequestTimeout(emitter.logger, cfg);

  if (!method) {
    throw new Error('Method is required');
  }

  const formattedMethod = methodsMap[method];

  if (!formattedMethod) {
    throw new Error(
      `Method "${method}" isn't one of the: ${Object.keys(methodsMap)}.`,
    );
  }

  const rateLimitDelay = getRateLimitDelay(emitter.logger, cfg);

  /*
   if cfg.followRedirect has value doNotFollowRedirects
   or cfg.followRedirect is not exists
   followRedirect option should be true
   */
  const requestOptions = {
    method: formattedMethod,
    uri: url,
    followRedirect,
    followAllRedirects: followRedirect,
    gzip: true,
    resolveWithFullResponse: true,
    simple: false,
    encoding: null,
    strictSSL: !cfg.noStrictSSL,
    timeout: requestTimeout,
  };

  const existingAuthHeader = (headers || []).find((header) => header._type === CREDS_HEADER_TYPE);

  switch (auth.type) {
    case authTypes.BASIC:
      if (existingAuthHeader) {
        existingAuthHeader.key = '';
      }

      headers.push({
        key: 'Authorization',
        // eslint-disable-next-line no-buffer-constructor
        value: `"Basic ${Buffer.from(`${auth.basic.username}:${auth.basic.password}`, 'utf8').toString('base64')}"`,
      });

      break;

    case authTypes.API_KEY:
      if (existingAuthHeader) {
        existingAuthHeader.key = '';
      }

      headers.push({
        key: auth.apiKey.headerName,
        value: `"${auth.apiKey.headerValue}"`,
      });

      break;
    case authTypes.OAUTH2:
      // eslint-disable-next-line no-case-declarations
      const token = await getValidToken();
      headers.push({
        key: 'Authorization',
        value: `"Bearer ${token}"`,
      });
      break;
    default:
      if (existingAuthHeader) {
        existingAuthHeader.key = '';
      }
  }

  if (headers && headers.length) {
    requestOptions.headers = headers.reduce(
      (headers, header) => {
        if (!header.key || !header.value) {
          return headers;
        }
        headers[header.key.toLowerCase()] = JsonataTransform.jsonataTransform(msg, { expression: header.value }, emitter);
        return headers;
      }, requestOptions.headers || {},
    );
  }

  return buildRequestBody()
    .then(() => {
      emitter.logger.debug('Got request body');
      return request(requestOptions);
    })
    .then(checkErrors)
    .then(processResponse)
    .then(async (result) => {
      emitter.logger.debug('Got request output');

      if (cfg.splitResult && Array.isArray(result)) {
      // Walk through chain of promises: https://stackoverflow.com/questions/30445543/execute-native-js-promise-in-series
      // eslint-disable-next-line no-restricted-syntax
        for (const item of result) {
          const output = messages.newMessageWithBody(item);
          output.attachments = msg.attachments;
          // eslint-disable-next-line no-await-in-loop
          await emitter.emit('data', output);
        }
      } else {
        const output = messages.newMessageWithBody(result);
        output.attachments = msg.attachments;
        await emitter.emit('data', output);
      }
      await rateLimit(emitter.logger, rateLimitDelay);
      await emitter.emit('end');
    })
    .catch(await buildErrorStructure);

  function checkOAuth2Keys(keys) {
    emitter.logger.debug('Check OAuth2 keys');
    if (!keys) {
      throw new Error('cfg.auth.oauth2.keys can not be empty');
    }
    if (!keys.access_token) {
      throw new Error('No access tokens were returned by the OAuth2 provider');
    }
    if (!keys.refresh_token) {
      throw new Error('No refresh tokens were returned by the OAuth2 provider. Try to add access_type:offline as an additional parameter');
    }
  }

  async function fetchNewToken() {
    emitter.logger.debug('Fetching new oauth2 token');
    const { oauth2 } = auth;
    const authTokenResponse = await request({
      uri: oauth2.tokenUri,
      method: 'POST',
      json: true,
      simple: false,
      resolveWithFullResponse: true,
      form: {
        refresh_token: oauth2.keys.refresh_token,
        grant_type: 'refresh_token',
        client_id: oauth2.clientId,
        client_secret: oauth2.clientSecret,
        scope: oauth2.scopes ? oauth2.scopes.join(' ') : '',
      },
    });

    emitter.logger.debug('New OAuth2 token fetched');

    if (authTokenResponse.statusCode >= 400) {
      throw new Error(`Error in authentication.  Status code: ${authTokenResponse.statusCode}, Body: ${JSON.stringify(authTokenResponse.body)}`);
    }

    return authTokenResponse.body;
  }

  async function getValidToken() {
    const { keys } = auth.oauth2;
    checkOAuth2Keys(keys);
    const tokenExpiryTime = new Date(keys.tokenExpiryTime);
    const now = new Date();
    if (now < tokenExpiryTime) {
      emitter.logger.debug('Previously valid token found');
      return keys.access_token;
    }

    const tokenRefreshStartTime = new Date();
    const oldRefreshToken = auth.oauth2.keys.refresh_token;
    auth.oauth2.keys = await fetchNewToken();
    if (!auth.oauth2.keys.refresh_token) {
      emitter.logger.debug('refresh_token was not returned, use old refresh_token');
      auth.oauth2.keys.refresh_token = oldRefreshToken;
    }
    checkOAuth2Keys(auth.oauth2.keys);
    // we need this check because expires_in is optional in oauth2
    if (auth.oauth2.keys.expires_in) {
      auth.oauth2.keys.tokenExpiryTime = (new Date(tokenRefreshStartTime.getTime()
        + (auth.oauth2.keys.expires_in * 1000))).toISOString();
    }
    emitter.logger.debug('Emit OAuth2 updateKeys');
    emitter.emit('updateKeys', { auth });

    return auth.oauth2.keys.access_token;
  }


  function buildRequestBody() {
    if (formattedMethod !== methodsMap.GET) {
      const bodyEncoding = {
        [contentTypes.FORM_DATA]: bodyEncodings.FORM_DATA,
        [contentTypes.URLENCODED]: bodyEncodings.URLENCODED,
      }[body.contentType] || bodyEncodings.RAW;

      // eslint-disable-next-line default-case
      switch (bodyEncoding) {
        case bodyEncodings.FORM_DATA:
          // eslint-disable-next-line no-case-declarations
          const existingContentTypeHeader = headers.find((header) => (
            // eslint-disable-next-line no-sequences
            header.key.match(/^content-type$/i),
            header.value === contentTypes.FORM_DATA
          ));

          if (existingContentTypeHeader) {
            existingContentTypeHeader.value = `"${formattedFormDataHeader}"`;
          } else {
            headers.push({
              key: 'Content-Type',
              value: `"${formattedFormDataHeader}"`,
            });
          }
          if (msg.attachments) {
            const attachments = Object.keys(msg.attachments).map(
              // eslint-disable-next-line no-unused-vars
              (key, index) => ({
                key,
                value: msg.attachments[key].url,
                filename: key,
                'Content-Type': msg.attachments[key]['content-type'],
              }),
            );

            // eslint-disable-next-line prefer-spread
            body.formData.push.apply(body.formData, attachments);
          }

          requestOptions.body = `--${bodyMultipartBoundary}`;

          return body.formData.reduce(
            (p, x) => p.then(() => processItem(x)), Promise.resolve(),
          ).then(() => {
            requestOptions.body = `${requestOptions.body}--`;
            return requestOptions.body;
          });

        case bodyEncodings.RAW:
          if (!body.raw) {
            break;
          }

          requestOptions.body = JsonataTransform.jsonataTransform(msg, { expression: body.raw }, emitter);
          if (typeof requestOptions.body === 'object') {
            requestOptions.body = JSON.stringify(requestOptions.body);
          }
          break;

        case bodyEncodings.URLENCODED:
          if (!body.urlencoded.length) {
            break;
          }

          // eslint-disable-next-line no-case-declarations
          const evaluatedUrlencoded = body.urlencoded.map((pair) => ({
            key: pair.key,
            value: JsonataTransform.jsonataTransform(msg, { expression: pair.value }, emitter),
          })).reduce((str, pair, index) => {
            const equation = `${encodeWWWFormParam(pair.key)}=${encodeWWWFormParam(pair.value)}`;

            return index === 0 ? equation : `${str}&${equation}`;
          }, null);

          requestOptions.body = evaluatedUrlencoded;
          break;
      }
      emitter.logger.trace('Got request body');
    }

    function processItem(item) {
      if (item.filename) {
        return request(item.value).then((result) => {
          requestOptions.body = `${requestOptions.body}\nContent-Disposition: form-data; name="${item.key}"; filename:"${item.filename}"\nContent-Type:${item['Content-Type']}\n\n${result}\n--${bodyMultipartBoundary}`;
        }).catch(() => {
          emitter.logger.trace('Item processing error');
        });
      }
      return Promise.resolve().then(() => {
        requestOptions.body = `${requestOptions.body}\nContent-Disposition: form-data; name="${item.key}"\n\n`
            + `${JsonataTransform.jsonataTransform(msg, { expression: item.value }, emitter)}\n--${bodyMultipartBoundary}`;
      });
    }

    return Promise.resolve(requestOptions.body);
  }

  async function buildErrorStructure(e) {
    if (cfg.enableRebound && (HTTP_ERROR_CODE_REBOUND.has(e.code)
      || e.message.includes('DNS lookup timeout'))) {
      emitter.logger.info('Component error');
      emitter.logger.info('Starting rebound');
      emitter.emit('rebound', e.message);
      emitter.emit('end');
    } else if (cfg.dontThrowErrorFlg) {
      const output = {
        errorCode: e.code,
        errorMessage: e.message,
        errorStack: e.stack,
      };
      await emitter.emit('data', messages.newMessageWithBody(output));
      await rateLimit(emitter.logger, rateLimitDelay);
      await emitter.emit('end');
    } else {
      emitter.logger.error('Component error');
      if (e.message === 'Error: ESOCKETTIMEDOUT') {
        e.message = `Timeout error! Waiting for response more than ${requestTimeout} ms`;
      }
      await emitter.emit('error', e);
      await rateLimit(emitter.logger, rateLimitDelay);
      await emitter.emit('end');
    }
  }

  /*
  * https://user-images.githubusercontent.com/13310949/41960520-9bd468ca-79f8-11e8-83f4-d9b2096deb6d.png
  * */
  function checkErrors(response) {
    const { statusCode } = response;
    emitter.logger.debug('Response statusCode %d', statusCode);
    if (statusCode >= 200 && statusCode < 300) {
      return Promise.resolve(response);
    }
    if (statusCode >= 300 && statusCode < 400) {
      if (followRedirect) {
        const REDIRECTION_ERROR = `${response.statusMessage
          || 'Redirection error.'} Please check "Follow redirect mode" if You want to use redirection in your request.`;
        if (cfg.dontThrowErrorFlg) {
          return Promise.resolve({
            statusCode,
            statusMessage: REDIRECTION_ERROR,
            headers: response.headers,
            body: response.body,
          });
        }
        const err = new Error(
          `Code: ${statusCode} Headers: ${JSON.stringify(
            response.headers,
          )} Body: ${JSON.stringify(
            response.body,
          )}. Error Message: ${REDIRECTION_ERROR}`,
        );
        err.code = statusCode;
        err.name = 'HTTP error';
        return Promise.reject(err);
      }
      return Promise.resolve(response);
    } if (statusCode >= 400 && statusCode < 1000) {
      if (cfg.dontThrowErrorFlg) {
        return Promise.resolve({
          headers: response.headers,
          body: response.body,
          statusCode,
          statusMessage: `${response.statusMessage || 'HTTP error.'}`,
        });
      }
      let errString = `Code: ${statusCode} Message: ${response.statusMessage || 'HTTP error'}`;
      let { body } = response;
      if (body && body.toString('utf8') !== '') {
        body = body.toString('utf8');
        errString = `${errString} Body: ${body}`;
      }
      const err = new Error(errString);
      err.code = statusCode;
      err.name = 'HTTP error';
      err.body = body;
      return Promise.reject(err);
    }
  }

  /*
  * parse response structure
  *
  * 1) If body is not exists return empty object {}
  * 2) If Content-type is exists in response try to parse by content type
  * 3) If Content-type is not exists try to parse as JSON. If we get parsing error
  * we should return response as is.
  *
  */
  function processResponse(response) {
    emitter.logger.trace('Got HTTP Response headers');
    emitter.logger.trace('Got HTTP Response body');

    if (response.body && response.body.byteLength === 0) {
      return Promise.resolve(buildResponseStructure({}));
    }

    const contType = response.headers['content-type'];

    emitter.logger.debug('Response content type: %o', contType);
    if (contType) {
      if (contType.includes('json')) {
        return Promise.resolve(response.body).then(JSON.parse).then(
          buildResponseStructure,
        );
      } if (contType.includes('xml')) {
        emitter.logger.info('Trying to parse response as XML');
        const parseOptions = {
          trim: false,
          normalize: false,
          explicitArray: false,
          normalizeTags: false,
          attrkey: '_attr',
          tagNameProcessors: [
            (name) => name.replace(':', '-'),
          ],
        };
        return xml2js(response.body, parseOptions)
          .then(buildResponseStructure).then((result) => {
            emitter.logger.info('Response successfully parsed');
            return result;
          });
      }
      if (contType.includes('image') || contType.includes('msword')
          || contType.includes('msexcel') || contType.includes('pdf')
          || contType.includes('csv') || contType.includes('octet-stream')
          || contType.includes('binary')) {
        return attachment.addAttachment.call(emitter, msg, `${uuidv1()}_${
          new Date().getTime()}`, response.body,
        response.headers['content-length'], contType)
          .then(() => {
            emitter.logger.info('Binary data successfully saved to attachments');
            return {};
          });
      }
      return Promise.resolve(
        buildResponseStructure(response.body.toString('utf8')),
      );
    }
    emitter.logger.info('Unknown content-type. Trying to parse as JSON');
    return Promise.resolve(response.body).then(JSON.parse)
      .then(buildResponseStructure)
      .catch(() => {
        emitter.logger.error('Parsing to JSON object is failed. Return response as is');
        return buildResponseStructure(response.body.toString('utf8'));
      });


    /*
    * return new output structure only if dontThrowErrorFlg is true
    *
    * New structure requirements:
    *
    * The outbound message body should include the HTTP response body from the REST call.
    * The message payload should include a headers section with
    * all of the headers received from the REST call.
    * The HTTP status code should also be included in the message payload.
    *
    * else return body of response
    */
    function buildResponseStructure(body) {
      if (typeof body !== 'object' && !Array.isArray(body)) {
        body = { result: body };
      }
      if (Array.isArray(body) && !cfg.splitResult) {
        body = { result: body };
      }
      if (cfg.dontThrowErrorFlg) {
        return {
          headers: response.headers,
          body,
          statusCode: response.statusCode,
          statusMessage: response.statusMessage,
        };
      }
      return body;
    }
  }
};
